src/encoding/encoding.rs

```diff
@@ -0,0 +1,235 @@
+                } else {
+                    color
+                };
+                self.draw_data
+                    .extend_from_slice(bytemuck::bytes_of(&DrawColor::new(color)));
+            }
+            BrushRef::LinearGradient(gradient) => {
+                self.add_ramp(&gradient.stops, alpha);
+                self.draw_tags.push(DrawTag::LINEAR_GRADIENT);
+                self.draw_data
+                    .extend_from_slice(bytemuck::bytes_of(&DrawLinearGradient {
+                        index: 0,
+                        p0: point_to_f32(gradient.start),
+                        p1: point_to_f32(gradient.end),
+                    }));
+            }
+            BrushRef::RadialGradient(gradient) => {
+                self.add_ramp(&gradient.stops, alpha);
+                self.draw_tags.push(DrawTag::RADIAL_GRADIENT);
+                self.draw_data
+                    .extend_from_slice(bytemuck::bytes_of(&DrawRadialGradient {
+                        index: 0,
+                        p0: point_to_f32(gradient.start_center),
+                        p1: point_to_f32(gradient.end_center),
+                        r0: gradient.start_radius,
+                        r1: gradient.end_radius,
+                    }));
+            }
+            BrushRef::SweepGradient(_gradient) => todo!("sweep gradients aren't done yet!"),
+        }
+    }
+
+    /// Encodes a begin clip command.
+    pub fn encode_begin_clip(&mut self, blend_mode: BlendMode, alpha: f32) {
+        use super::DrawBeginClip;
+        self.draw_tags.push(DrawTag::BEGIN_CLIP);
+        self.draw_data
+            .extend_from_slice(bytemuck::bytes_of(&DrawBeginClip::new(blend_mode, alpha)));
+        self.n_clips += 1;
+    }
+
+    /// Encodes an end clip command.
+    pub fn encode_end_clip(&mut self) {
+        self.draw_tags.push(DrawTag::END_CLIP);
+        // This is a dummy path, and will go away with the new clip impl.
+        self.path_tags.push(PathTag::PATH);
+        self.n_paths += 1;
+        self.n_clips += 1;
+    }
+
+    // Swap the last two tags in the path tag stream; used for transformed
+    // gradients.
+    pub fn swap_last_path_tags(&mut self) {
+        let len = self.path_tags.len();
+        self.path_tags.swap(len - 1, len - 2);
+    }
+
+    fn add_ramp(&mut self, stops: &[ColorStop], alpha: f32) {
+        let offset = self.draw_data.len();
+        let stops_start = self.color_stops.len();
+        if alpha != 1.0 {
+            self.color_stops.extend(stops.iter().map(|s| ColorStop {
+                offset: s.offset,
+                color: color_with_alpha(s.color, alpha),
+            }));
+        } else {
+            self.color_stops.extend_from_slice(stops);
+        }
+        self.patches.push(Patch::Ramp {
+            offset,
+            stops: stops_start..self.color_stops.len(),
+        });
+    }
+}
+
+fn color_with_alpha(mut color: Color, alpha: f32) -> Color {
```

Comment:
This feels like it should be a method on Color?

Reply:
Yes, I’ll update add it to my pending peniko patch. For context, lottie has an additional opacity modifier and I wanted to be able to encode this without rebuilding brushes.